### PR TITLE
Added CodeCov coverage report support

### LINKED ISSUE #284: Add CodeCov
⚠️ This issue is generated, it means the examples and the namings do not necessarily correspond to the language of this repository.
Also, if you are a maintainer, please add any clarification and instructions about this issue.

Sorry if this is already wholly/partially implemented. Feel free to let me know about the state of this issue in the repo.

Related to https://github.com/meilisearch/integration-guides/issues/59

----

Codecov helps us to know if a PR add non-tested code. If it's the case, we don't accept the PR. 
It's easy to install, easy to configure.

### TODO:

- [ ] Add CodeCov to this repository.
- [ ] Add the CodeCov badge on this repository README.md
- [ ] Ensure a minimum coverage is accepted (optional)

### PR DESCRIPTION
**Pull request by connorhu**
# Pull Request

## Related issue
Fixes #284

## What does this PR do?
- Integrates CodeCov

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: .github/workflows/tests.yml
Comment (norkunas) [inline .github/workflows/tests.yml:72]: Thanks, but imho it should run on the same testing step instead of testing twice one without coverage and second with. 
Also I think `coverage-text` would be better option unless you plan to add upload step somewhere?

Comment (connorhu) [reply .github/workflows/tests.yml:72]: I haven't added the upload yet because I'm wondering if it makes sense to generate a coverage report for each test and upload it in one go. Something like this:

https://github.com/doctrine/orm/blob/bfb033fe3c8f567e088802fc3641e6bdd163f7ce/.github/workflows/continuous-integration.yml#L93-L97

and

https://github.com/doctrine/orm/blob/bfb033fe3c8f567e088802fc3641e6bdd163f7ce/.github/workflows/continuous-integration.yml#L319-L342

File: .github/workflows/tests.yml
Comment (norkunas) [inline .github/workflows/tests.yml:117]:
```suggestion
    runs-on: 'ubuntu-latest'
```

**Review by norkunas — CHANGES_REQUESTED** [review]
For consistency use single quotes

File: .github/workflows/tests.yml
Comment (connorhu) [reply .github/workflows/tests.yml]: In the name of consistency should I reformat the quotes of "Remove doctrine/annotations"?

File: .github/workflows/tests.yml
Comment (norkunas) [inline .github/workflows/tests.yml:115]:
```suggestion
  upload-coverage:
```

File: .github/workflows/tests.yml
Comment (norkunas) [inline .github/workflows/tests.yml:127]:
```suggestion
        uses: actions/download-artifact@v4
```

File: .github/workflows/tests.yml
Comment (norkunas) [inline .github/workflows/tests.yml:132]:
```suggestion
        uses: codecov/codecov-action@v4
```

**Comment by connorhu** [comment]
We have to lock phpunit version because coverage report is broken with PHP 7.4, Symfony 5.4 and phpunit 8. With this constellation, code coverage did not work. One solution is to lock the phpunit version, another is to create a .ci/github/phpunit/phpunit.xml.dist file and lock the version only in that file. Both solutions have their advantages and disadvantages.

```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to SebastianBergmann\CodeCoverage\CodeCoverage::__construct() must be an instance of SebastianBergmann\CodeCoverage\Driver\Driver, null given, called in /home/runner/work/meilisearch-symfony/meilisearch-symfony/vendor/bin/.phpunit/phpunit-8.5-0/src/TextUI/TestRunner.php on line 526 and defined in /home/runner/work/meilisearch-symfony/meilisearch-symfony/vendor/phpunit/php-code-coverage/src/CodeCoverage.php:122
Stack trace:
#0 /home/runner/work/meilisearch-symfony/meilisearch-symfony/vendor/bin/.phpunit/phpunit-8.5-0/src/TextUI/TestRunner.php(526): SebastianBergmann\CodeCoverage\CodeCoverage->__construct()
#1 /home/runner/work/meilisearch-symfony/meilisearch-symfony/vendor/bin/.phpunit/phpunit-8.5-0/src/TextUI/Command.php(235): PHPUnit\TextUI\TestRunner->doRun()
#2 /home/runner/work/meilisearch-symfony/meilisearch-symfony/vendor/bin/.phpunit/phpunit-8.5-0/src/TextUI/Command.php(194): PHPUnit\TextUI\Command->run()
#3 /home/runner/work/meilisearch-symfony/meilisearch-symfony in /home/runner/work/meilisearch-symfony/meilisearch-symfony/vendor/phpunit/php-code-coverage/src/CodeCoverage.php on line 122
Script SYMFONY_DEPRECATIONS_HELPER='ignoreFile=./tests/baseline-ignore' ./vendor/bin/simple-phpunit --colors=always --verbose handling the test:unit event returned with error code 255
```

**Comment by norkunas** [comment]
I don't mind locking it as long as tests runs and are green :)

**Comment by norkunas** [comment]
The CI breaks now because bors expects these jobs:
```
status = [
    'integration-tests (PHP 7.4) (Symfony 5.4.*)',
    'integration-tests (PHP 8.0) (Symfony 6.0.*)',
    'integration-tests (PHP 8.1) (Symfony 6.1.*)',
    'integration-tests (PHP 8.2) (Symfony 6.2.*)',
    'integration-tests (PHP 8.3) (Symfony 6.3.*)',
    'integration-tests (PHP 8.3) (Symfony 6.4.*)',
    'integration-tests (PHP 8.3) (Symfony 7.0.*)',
]
```

So I think either restore using with `.*` and then just strip it before passing to codecov to avoid limitations as you said (will be less changes) or update the bors.toml

**Comment by connorhu** [comment]
Both solutions are good. If editing bors.toml causes further problems, I'll put an wildcard in the name of the step instead.

**Comment by norkunas** [comment]
The less changes are better :)

**Comment by norkunas** [comment]
```
Error: No coverage reports found. Please make sure you're generating reports successfully.
Warning: Codecov: 
                      Failed to properly upload report: The process '/home/runner/work/_actions/codecov/codecov-action/v4/dist/codecov' failed with exit code 1
```

**Comment by connorhu** [comment]
Now it looks like it will find it:
```
[2024-02-14T08:05:45.455Z] ['info'] => Found 5 possible coverage files:
  phpunit-7.4-5.4-coverage/coverage.xml
  phpunit-8.0-6.0-coverage/coverage.xml
  phpunit-8.1-6.1-coverage/coverage.xml
  phpunit-8.2-6.2-coverage/coverage.xml
  phpunit-8.3-7.0-coverage/coverage.xml
```

**Comment by norkunas** [comment]
Why it's not working with v4? Adding a new feature should use last versions :thinking:

**Comment by connorhu** [comment]
> Why it's not working with v4? Adding a new feature should use last versions 🤔

The problem was caused by the fact that I only transferred to v4 on download, not on upload. v3 uploaded, v4 downloaded.

**Comment by connorhu** [comment]
Shiny 86% reported. Nice.
report

**Comment by norkunas** [comment]
I still suggest to revert changes about wildcard, because in repo settings jobs are made as required with previous names, and I don't have permissions to do that

**Review by norkunas — APPROVED** [review]
Good job :tada:

After #323 CI will be green (I can't review/merge my own PR's)
Before merging this we should probably wait for @brunoocasali approval.

**Comment by connorhu** [comment]
I signed up for that PR. If it went in I'll rebase this and we'll green it.

**Comment by connorhu** [comment]
About this php-cs-fixer problem: I think you should hardcode the php-cs-fixer version everywhere, and apply the fixer changes with version changes. Elsewhere I just ran into that two versions interpreting something completely differently (incorrectly) and it took me some time to find out what the source of the problem was.

**Comment by norkunas** [comment]
That's just because `@Symfony` ruleset rules has changed, it doesn't change too much, so therefore I'm not sure about locking cs fixer

**Comment by norkunas** [comment]
@connorhu you can rebase now :)

**Review by norkunas — APPROVED** [review]

**Comment by norkunas** [comment]
bors merge

**Comment by norkunas** [comment]
bors merge

**Comment by norkunas** [comment]
bors cancel

**Comment by norkunas** [comment]
@connorhu could you squash your commits?

**Comment by connorhu** [comment]
@norkunas of course. done

**Comment by norkunas** [comment]
Thank you @connorhu

**Pull request merged**
